### ISSUE TITLE
Pattern matcher calls ExecutionOutputLink with non-grounded(free) variable

### ISSUE THREAD

**Issue by noskill** [opencog/atomspace#1908]
Here is an example:

```scheme

(add-to-load-path ".")

(use-modules (opencog))
(use-modules (opencog query))
(use-modules (opencog exec))


(Inheritance (Concept "red") (Concept "color"))
(Inheritance (Concept "blue") (Concept "color"))
(Inheritance (Concept "green") (Concept "color"))


(define (schema2 A B)
    (display "Executing schema2 with ")
    (display (list (cog-name A) (cog-name B)))
    (newline)
    (if (equal? (cog-name A) (cog-name B))
        (Concept "Same color")
        (Concept "Different color")))


(define (classify A)
    (display "Executing predicate with ")
    (display (cog-name A))
    (newline)
    (if (equal? (cog-name A) "Same color")
        (stv 1 1)
        (stv 0 1)))

        
(define bind-link
    (BindLink
        (VariableList
            (Variable "$X")
            (Variable "$Y"))
        (And
            (Inheritance
                (Variable "$X")
                (Concept "color"))
            (Inheritance
                (Variable "$Y")
                (Concept "color"))
            (Evaluation
                (GroundedPredicate "scm: classify")
                (ExecutionOutput
                    (GroundedSchema "scm: schema2")
                    (List
                         (Variable "$X")
                         (Variable "$Y"))
                )
            ))
        (List (Variable "$X") (Variable "$Y")))
        )

(define result (cog-execute! bind-link))
```

Running this code gives in output lines like these:
```

Executing schema2 with ($Y green)
Executing schema2 with ($Y $X)
```
So function schema2 gets called with non-grounded variable. 
Documentation somewhat ambiguous  https://wiki.opencog.org/w/SatisfactionLink_and_BindLink#Variable_Binding, but it seems that variables should become bounded in pattern search.

**Comment by linas** [comment]
There is no a-priori reason that one variable might not be grounded by another; this is even needed and expected by some of the test cases.  So this might seem like unexpected behavior, but it is "working as designed".

If you want to prevent this from happening, you can tighten the search space with a declaration:
```
(VariableList
            (TypedVariable  (Variable "$X") (Type 'ConceptNode))
             (TypedVariable  (Variable "$Y") (Type 'ConceptNode)))
``` 
and then the Variables  will never appear.

**Comment by linas** [comment]
By the way, it is much more efficient to just say `(if (equal? A B)` instead of `(if (equal? (cog-name A) (cog-name B))`  because it is impossible to have two different nodes with the same name (in the same atomspace).

**Comment by noskill** [comment]
Ok I understand that generally variables may be grounded by variables, but in this case (Inheritance (Variable "$X") (Concept "color")) does not hold, so it doesn't worth to call GroundedSchema with (Variable "$X"), does it?

**Comment by ngeiswei** [comment]
I believe it does hold, as X matches with Y, and Y matches with X, right?

**Comment by noskill** [comment]
@ngeiswei Yes, X matches Y, but neither of them inherits from (Concept "color"). I was expecting pattern matcher to evaluate all the inheritance constraints first for performance reasons, because in this issue https://github.com/opencog/atomspace/issues/1907 there is no TypedVariable and still grounded predicate isn't called with variable nodes.

**Comment by linas** [comment]
In the middle of what you posted up top, you wrote:
```
           (Inheritance
                (Variable "$X")
                (Concept "color"))
            (Inheritance
                (Variable "$Y")
                (Concept "color"))
```
So both inherit from color.

**Comment by ngeiswei** [comment]
@noskill said:
> because in this issue #1907 there is no TypedVariable and still grounded predicate isn't called with variable nodes

That is only due to discarding self-matches. The 2 inheritances are between different concepts (`Size` and `Items`) so there's no way to have their variables match with one another.

If typing the variable is inconvenient because the variables don't have obvious types except for not being variables, then I wonder... @linas, is there a form of type negation? Something that would allow to write

```scheme
(TypedVariable
  (Variable "$X")
  (Not (Type "VariableNode")))
```

?

This would turn the type system into a sigma algebra rather than an a topology (ahah, just kidding, trying to sound like you @linas, though I think I might be on spot, maybe).

**Comment by ngeiswei** [comment]
Thanks for recalling these type links, there are really cool. In fact `TypeSetLink` represents the type intersection operator the unifier incompletely implements, maybe there's some possible reuse here.
